horovod/common/response_cache.cc

```diff
@@ -0,0 +1,223 @@
+
+ResponseCache::CacheState ResponseCache::cached(const Request& message) const {
+  auto it = tensor_name_to_bit_.find(message.tensor_name());
+  if (it != tensor_name_to_bit_.end()) {
+    // If entry associated with this request already exists in cache, check
+    // if tensor parameters match. If not, return that entry is invalid.
+    uint32_t cache_bit = it->second;
+    auto& cache_params = std::get<1>(*cache_iters_[cache_bit]);
+    return (cache_params.device == message.device() &&
+            cache_params.dtype == message.tensor_type() &&
+            cache_params.shape == message.tensor_shape())
+               ? CacheState::HIT
+               : CacheState::INVALID;
+  } else {
+    return CacheState::MISS;
+  }
+}
+
+ResponseCache::CacheState
+ResponseCache::cached(const Response& response,
+                      const TensorParams& params) const {
+  assert(response.tensor_names().size() == 1);
+  auto it = tensor_name_to_bit_.find(response.tensor_names()[0]);
+  if (it != tensor_name_to_bit_.end()) {
+    uint32_t cache_bit = it->second;
+    auto& cache_params = std::get<1>(*cache_iters_[cache_bit]);
+    return (cache_params.device == params.device &&
+            cache_params.dtype == params.dtype &&
+            cache_params.shape == params.shape)
+               ? CacheState::HIT
+               : CacheState::INVALID;
+  } else {
+    return CacheState::MISS;
+  }
+}
+
+void ResponseCache::put_(const Response& response, TensorParams& params) {
+  uint32_t cache_bit;
+  auto cache_state = this->cached(response, params);
+
+  // Disallow caching name-conflicted responses here. Invalid cache entries
+  // must be removed prior to caching new entries.
+  if (cache_state == CacheState::INVALID) {
+    throw std::logic_error(
+        "Trying to overwrite cached response with existing name. "
+        "This is not allowed.");
+  }
+
+  if (this->cached(response, params) == CacheState::HIT) {
+    cache_bit = tensor_name_to_bit_[response.tensor_names()[0]];
+    auto it = cache_iters_[cache_bit];
+    cache_.push_front(std::move(*it));
+    cache_.erase(it);
+  } else if (cache_.size() == capacity_) {
+    auto& entry = cache_.back().first;
+    cache_bit = tensor_name_to_bit_[entry.tensor_names()[0]];
+    tensor_name_to_bit_.erase(entry.tensor_names()[0]);
+    cache_.pop_back();
+    cache_.push_front(std::make_pair(response, std::move(params)));
+  } else {
+    cache_bit = cache_iters_.size();
+    cache_iters_.resize(cache_bit + 1);
+    cache_.push_front(std::make_pair(response, std::move(params)));
+  }
+
+  cache_iters_[cache_bit] = cache_.begin();
+  tensor_name_to_bit_[response.tensor_names()[0]] = cache_bit;
+
+  bits_outdated_ = true;
+}
+
+void ResponseCache::put(const Response& response, TensorTable& tensor_table) {
+  if (capacity_ == 0) {
+    return;
+  }
+
+  // If response is fused, split back into individual responses
+  if (response.tensor_names().size() > 0) {
```

Comment:
CLion suggests to replace all `.size() > 0` with `!.empty()`

Reply:
The original statement here was actually incorrect. It should've been `response.tensor_names().size() > 1`. Corrected it.